services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala

```diff
@@ -13,38 +13,63 @@ import scala.concurrent.{ ExecutionContext, Future }
 class GitLabProjectVersioning(httpClient: HttpClient, config: GitLabConfig)
     extends ProjectVersioning[VersioningException] {
 
-  override def updateFile(project: Project, projectFile: ProjectFile, version: Option[Version])(
+  override def updateFile(project: Project, projectFile: ProjectFile, version: Option[PipelineVersion])(
     implicit ec: ExecutionContext
   ): AsyncResult[String] = {
     val path = URLEncoder.encode(projectFile.path.toString, "UTF-8")
-    val fileUrl = s"${config.url}projects/${project.repository}/repository/files/$path"
-    val versionValue = version.map(_.name).getOrElse(config.defaultFileVersion)
+    val repositoryId: Repository = project.repository match {
+      case Some(repository) => repository
+      case None             => throw VersioningException(s"No repository for project: $project")
+    }
+    val fileUrl = s"${config.url}projects/${repositoryId.value}/repository/files/$path"
 
+    getProjectVersions(project).flatMap(versions => {
+      val latestVersion: Either[VersioningException, PipelineVersion] = (version, versions) match {
+        case (Some(_), Left(error))                             => Left(error)
+        case (Some(tagName), Right(v :: _)) if v.name < tagName => Right(tagName)
+        case (Some(tagName), Right(v :: _)) if v.name >= tagName =>
+          Left(VersioningException(s"Your version $tagName is out of date. Current version of project: ${v.name}"))
+        case (Some(_), Right(_))   => Right(PipelineVersion(config.defaultFileVersion))
+        case (None, Right(v :: _)) => Right(v.name)
+        case (None, Right(Nil))    => Right(PipelineVersion(config.defaultFileVersion))
+        case (None, Left(error))   => Left(error)
+      }
+      latestVersion match {
+        case Left(error) =>
+          Future.successful(Left(error))
+        case Right(tagName) =>
+          val payload =
+            Json.stringify(Json.toJson(UpdateFileRequest(projectFile.content, tagName.toString, config.defaultBranch)))
+          for {
+            updateResult <- httpClient.put(fileUrl, payload = payload, headers = config.token)
+            createResult <- httpClient.post(fileUrl, payload = payload, headers = config.token)
```

Comment:
What?

Reply:
Why?

Reply:
I thought that it will look more accurate than chain of try to updateFile and if this file doesn't exist, createFile, but will fix it, no problem